chtest.c

```diff
@@ -0,0 +1,260 @@
+#include <linux/moduleparam.h>		//library	
+#include <linux/init.h>           // Macros used to mark up functions e.g. __init __exit
+#include <linux/module.h>         // Core header for loading LKMs into the kernel
+#include <linux/device.h>         // Header to support the kernel Driver Model
+#include <linux/kernel.h>         // Contains types, macros, functions for the kernel
+#include <linux/fs.h>             // Header for the Linux file system support
+#include <linux/uaccess.h>          // Required for the copy to user function
+#define  DEVICE_NAME "chtest"    ///< The device will appear at /dev/chtest using this value
+#define  CLASS_NAME  "Test_Class"        ///< The device class -- this is a character device driver
+#define BSIZE  1024
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Vitaliy Izhyk <vizhyk1@gmail.com");
+MODULE_DESCRIPTION("Char_Driver");
+MODULE_VERSION("0.1");
+
+static int B_size = 0;			///<BUFFER from parameter
+module_param(B_size, int, S_IRUGO);
+char *new_buff;
+static int clrflag;				//flag for sys
+static int    majorNumber;                  ///< Stores the device number -- determined automatically
+static char   message[BSIZE] = {0};           ///< Memory for the string that is passed from param
+static short  size_of_message;              ///< Used to remember the size of the string stored
+static int    numberOpens = 0;              ///< Counts the number of times the device is opened
+static struct class*  Driver_Class  = NULL; ///< The device-driver class struct pointer
+static struct device* Driver_Device = NULL; ///< The device-driver device struct pointer
+static struct proc_dir_entry *procptr;
+
+					//Prototypes
+static ssize_t procwrite(struct file *file, const char __user *ubuf,size_t count, loff_t *ppos); //proc prototype
+static ssize_t procread(struct file *file, char __user *ubuf,size_t count, loff_t *ppos);//proc prototype
+
+static ssize_t clrflag_show(struct kobject *kobj, struct kobj_attribute *chtest_kobject, char *buf); //sys prototype
+static ssize_t clrflag_store(struct kobject *kobj,struct kobj_attribute *chtest_kobject, char *buf, size_t count); //sys prototype
+
+static int     dev_open(struct inode *, struct file *);
+static int     dev_release(struct inode *, struct file *);
+static ssize_t dev_read(struct file *, char *, size_t, loff_t *);
+static ssize_t dev_write(struct file *, const char *, size_t, loff_t *);
+
+		//device struct
+static struct file_operations fops =
+ {
+	.open = dev_open,
+	.read = dev_read,
+	.write = dev_write,
+	.release = dev_release,
+	.owner = THIS_MODULE,
+ };
+		//proc struct
+static struct file_operations procops = 
+ {
+	.owner = THIS_MODULE,
+	.read = procread,
+	.write = procwrite,
+ };
+		//sys struct
+static struct kobj_attribute sys_attribute =__ATTR(clrflag, 0666, clrflag_show,clrflag_store);
+
+static int __init chtest_init(void)
+ {
+   printk(KERN_INFO "chtest: Initializing the chtest LKM\n");
+if (B_size > 0) 
```

Comment:
Your indentation below this point is awful. It's almost impossible to read this code.